### PR TITLE
Add metrics for vault

### PR DESCRIPTION
**Pull request by waiariki-koia**
Have added implementation to push metrics to the push gateway.

Added times of successes and errors.
Added error count
Added time until credentials expire

### CODE REVIEW

**Review by rhysemmas — DISMISSED** [review]

**Comment by Joseph-Irving** [comment]
what happens if someone doesn't set the prom gateway address or doesn't want to?

**Comment by waiariki-koia** [comment]
If the gateway address is not defined, it becomes an empty string by default. There is a check in the `Push()` method to see if the address is set or not. If the address is no set no metrics get pushed

File: pkg/metrics/pushgateway.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+	errorCount = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: promNamespace,
+		Name:      "error_count",
+		Help:      "Number of errors when renewing credentials",
+	})
+
+	successTime = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: promNamespace,
+		Name:      "last_renewal_success_timestamp_seconds",
```
Comment (Joseph-Irving) [inline pkg/metrics/pushgateway.go]: is this in seconds or is it a timestamp?

File: pkg/metrics/pushgateway.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+	errorTime = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: promNamespace,
+		Name:      "last_renewal_error_timestamp_seconds",
+		Help:      "The timestamp of the last error during renewal of a secret",
+	})
+
+	errorCount = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: promNamespace,
+		Name:      "error_count",
```
Comment (Joseph-Irving) [inline pkg/metrics/pushgateway.go]: `credential_renewal_errors_total` would be more in line with prom naming conventions

File: pkg/metrics/pushgateway.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+	successTime = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: promNamespace,
+		Name:      "last_renewal_success_timestamp_seconds",
+		Help:      "The timestamp of the last successful renewal of a secret",
+	})
+
+	leaseExpiration = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: promNamespace,
+		Name:      "time_until_secrets_expire",
```
Comment (Joseph-Irving) [inline pkg/metrics/pushgateway.go]: `expiry_time_seconds` or something like that would be more prometheusy 
see https://prometheus.io/docs/practices/naming/

File: pkg/metrics/pushgateway.go (hunk 1 above)
Comment (waiariki-koia) [reply pkg/metrics/pushgateway.go]: This is a unix timestamp

**Comment by Joseph-Irving** [comment]
looks good, do you want to add something to the readme to document the metrics and why we've opted for the push gateway model here

**Comment by waiariki-koia** [comment]
Yeah will do. 

Also wondered if it's useful/if its even possible to add more labels to make querying easier. 

At the moment we only have pod name, namespace and cluster, the normal kube labels we get come from the gateway as its the one getting scraped.

I'm not sure how we would be able to get something like deployment name or other grouping type things though

**Comment by Joseph-Irving** [comment]
I think that seems like enough info to begin with, if we feel we need more we could add it later

**Comment by waiariki-koia** [comment]
https://thanos.usw.cloud/graph?g0.range_input=30m&g0.max_source_resolution=0s&g0.expr=vault_creds_credential_expiry_time_seconds&g0.tab=0 to see metrics being removed

**Review by rhysemmas — APPROVED** [review]

**Pull request merged**